### ISSUE TITLE
Strange out of order root nodes with MP_Node

### ISSUE THREAD

**Issue by zbyte64** [django-treebeard/django-treebeard#65]
Have an issue where the system seems to retrieve the wrong last root node causing the path increment logic to create a conflicting path.

```
>>> Concept.find_problems()
([], [], [], [], [])
>>> [(o, o.path) for o in Concept.get_root_nodes()]
[(<Concept: Law>, '0003'), (<Concept: jurisdiction>, '0002'), (<Concept: identity document>, '0004'), (<Concept: currency>, '0001')]
>>> Concept.get_last_root_node()._inc_path()
'0004' #this path already exists!
>>> pprint(Concept.dump_bulk())
[{'data': OrderedDict([('title', 'Law'),
                       ('slug', 'law'),
                       ('definition',
                        'a system of rules and guidelines which are enforced '
                        'through social institutions to govern behavior'),
                       ('example', ''),
                       ('url_path', 'law'),
  'id': 25},
 {'data': OrderedDict([('title', 'jurisdiction'),
                       ('slug', 'jurisdiction'),
                       ('definition',
                        'the limits or territory within which authority may be '
                        'exercised; the power, right, or authority to '
                        'interpret and apply the law'),
                       ('example', ''),
                       ('url_path', 'jurisdiction'),
  'id': 29},
 {'data': OrderedDict([('title', 'identity document'),
                       ('slug', 'identity-document'),
                       ('definition',
                        'any legal document which may be used to verify '
                        "aspects of a person's identity"),
                       ('example', ''),
                       ('url_path', 'identity-document'),
  'id': 34},
 {'data': OrderedDict([('title', 'currency'),
                       ('slug', 'currency'),
                       ('definition',
                        'medium of exchange value, defined by reference to the '
                        'geographical location of the authorities responsible '
                        'for it'),
                       ('example', ''),
                       ('url_path', 'currency'),
  'id': 23}]
>>> Concept.fix_tree(destructive=True)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.IntegrityError: duplicate key value violates unique constraint "focus_concept_path_key"
DETAIL:  Key (path)=(0002) already exists.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/treebeard/mp_tree.py", line 773, in fix_tree
    cls.load_bulk(dump, None, True)
  File "/usr/local/lib/python3.6/site-packages/treebeard/models.py", line 117, in load_bulk
    node_obj = cls.add_root(**node_data)
  File "/usr/local/lib/python3.6/site-packages/treebeard/mp_tree.py", line 625, in add_root
    return MP_AddRootHandler(cls, **kwargs).process()
  File "/usr/local/lib/python3.6/site-packages/treebeard/mp_tree.py", line 345, in process
    newobj.save()
  File "/code/apps/focus/models.py", line 149, in save
    return super(Concept, self).save(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 796, in save
    force_update=force_update, update_fields=update_fields)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 824, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 908, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/base.py", line 947, in _do_insert
    using=using, raw=raw)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 1045, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1054, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/usr/local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: duplicate key value violates unique constraint "focus_concept_path_key"
DETAIL:  Key (path)=(0002) already exists.

```

**Comment by zbyte64** [comment]
Seems that `get_root_nodes` is not reversible somehow:

```
>>> Concept.get_root_nodes()
<QuerySet [<Concept: Law>, <Concept: jurisdiction>]>
>>> Concept.get_root_nodes().reverse()
<QuerySet [<Concept: Law>, <Concept: jurisdiction>]>
```

That would definitely explain the error. Now to find out why. Using django 1.10.6

**Comment by zbyte64** [comment]
`MP_Node` is not proving a default ordering, thus `reverse` does nothing.

**Comment by philippeluickx** [comment]
Related? https://github.com/django-treebeard/django-treebeard/issues/81

**Comment by jrief** [comment]
closed via #112
